[resourcedetectionprocessor] Initial system detector

mx-psi opened the pull request:
**Description:** <Describe what has changed.>
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->

Adds resource detection of hostname and host information when running on a bare host.

**Link to tracking Issue:** #1292 

**Testing:** Unit tests were added; processor was tested on a Linux and macOS host.

**Documentation:** 
- Updated README with the information gathered by this detector
- Updated examples

mx-psi committed:
Add system resource detector

mx-psi committed:
Add unit tests

jrcamp commented on processor/resourcedetectionprocessor/internal/system/system.go:
```diff
@@ -0,0 +1,72 @@
+		attrs.InsertString(conventions.AttributeHostName, fqdn)
+	}
+
+	attrs.InsertString(conventions.AttributeHostHostname, hostname)
```
host.hostname was removed from spec. I filed issue just now to remove throughout (https://github.com/open-telemetry/opentelemetry-collector/issues/2029). Don't think I would add it here now.

If we need to collect both hostname and the fqdn we may need to propose adding a `host.fqdn` semantic convention. Or if we could somehow make the detectors configurable and fqdn could be an option that might work too. But I think that would require refactoring the config for detectors.

mx-psi replied:
Oh, I didn't realize, thanks for pointing it out. An alternative that I considered for this was `net.host.name` but I am not sure if it would be really equivalent.

jrcamp replied:
Looks like net.host.name is part of [trace semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/9a803247d8fb652f9cf7b11270820415ed5dd2ce/specification/trace/semantic_conventions/span-general.md) though, not resource conventions so won't work here.

> Otherwise, maybe it would be most useful to use the FQDN as host.name?

Sounds good.

jrcamp commented on processor/resourcedetectionprocessor/internal/system/metadata_others.go:
```diff
@@ -0,0 +1,35 @@
+)
+
+func (*systemMetadataImpl) FQDNAvailable() bool {
+	_, err := os.Stat("/bin/hostname")
```
This wouldn't work in our Docker images because we don't include a base OS with utilities like hostname. But we also don't include /etc/hosts, etc. so I'm not sure we'd have a fqdn anyway. 🤔 

However maybe we should use https://github.com/Showmax/go-fqdn since it has Windows support (as well as not relying on hostname binary).

@keitwb any ideas?

mx-psi replied:
> This wouldn't work in our Docker images because we don't include a base OS with utilities like hostname. But we also don't include /etc/hosts, etc. so I'm not sure we'd have a fqdn anyway. 🤔

I am aware of this  (you would have to install the `net-tools` package on the image) but I was thinking of other use cases (e.g. users that rely on `.deb` or `.rpm` packages). The project does plan to support the latter, right?

> However maybe we should use https://github.com/Showmax/go-fqdn since it has Windows support (as well as not relying on hostname binary).

That sounds like a good idea to me.

jrcamp replied:
Yep we publish deb and rpm so it would be available in some cases. But ideally we would want the processor to behave the same across as many environments as possible.

mx-psi replied:
👍 I am going to use `go-fqdn` then if @keitwb does not have any other ideas, which would have more widespread support and seems reasonable

mx-psi commented on processor/resourcedetectionprocessor/internal/system/system.go:
```diff
@@ -0,0 +1,72 @@
+	}
+
+	attrs.InsertString(conventions.AttributeHostHostname, hostname)
+	attrs.InsertString(conventions.AttributeHostType, hostType)
```
So I re-read the spec (turns out I was reading and old version thanks to Github search) and I think it makes more sense to remove this and instead have `os.type` mapped to `runtime.GOOS`. 

If there is interest at some point we could pursue a `os.architecture` resource convention that can be added here, but for now I think it's preferable to have `os.type`.

mx-psi committed:
Get FQDN through go-fqdn

mx-psi committed:
Change from `host.type` to `os.type`

mx-psi committed:
Remove `host.hostname` since it was removed from the spec

mx-psi committed:
Improve error messages

mx-psi commented:
@jrcamp I addressed your comments and did some minor improvements. Mind having a second look? For context I

- removed `host.hostname` and set `host.name` to the FQDN, now obtained using `go-fqdn` and
- change `host.type` to `os.type` and set it to uppercase `runtime.GOOS` (see [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/resource/semantic_conventions/os.md))

mx-psi committed:
Merge branch 'master' into mx-psi/system-resource

mx-psi commented:
Friendly ping @jrcamp

jrcamp reviewed (approved):
lgtm, thanks!
